src/property.ts

```diff
@@ -4,252 +4,188 @@
+                        prop.choices.join(', ')
+                    }]`,
                     field: prop.name,
-                    value: castValue,
                 });
             }
         }
+        if (prop.check && !prop.check(castValue)) {
+            throw new ValidationError({
+                message: `Violated check constraint of ${prop.name}${prop.check.name
+                    ? ` (${prop.check.name})`
+                    : ''
+                }`,
+                field: prop.name,
+                value: castValue,
+            });
+        }
+    }
 
-        // check minItems and maxItems
-        if (prop.minItems && result.length < prop.minItems) {
-            throw new ValidationError(`Violated the minItems constraint of ${prop.name}. Less than the required number of elements (${result.length} < ${prop.minItems})`);
+    // check minItems and maxItems
+    if (prop.minItems && result.length < prop.minItems) {
+        throw new ValidationError(`Violated the minItems constraint of ${prop.name}. Less than the required number of elements (${result.length} < ${prop.minItems})`);
+    }
+    if ((prop.maxItems || prop.maxItems === 0) && result.length > prop.maxItems) {
+        throw new ValidationError(`Violated the maxItems constraint of ${prop.name}. More than the allowed number of elements (${result.length} > ${prop.maxItems})`);
+    }
+    return inputValue instanceof Array
+        ? result
+        : result[0];
+};
+
+export const createPropertyDefinition = (opt: PropertyDefinitionInput): PropertyDefinition => {
+    const {
+        type: inputType,
+        default: inputDefault,
+        ...rest
+    } = opt;
+    const defaultType = ((opt.min !== undefined || opt.max !== undefined) && !opt.type)
+        ? 'integer'
+        : 'string';
+    const type: DbType = inputType || defaultType;
+
+    let defaultCast;
+
+    if (!opt.cast) { // set the default util.cast functions
+        if (type === 'integer') {
+            defaultCast = util.castInteger;
+        } else if (type === 'string') {
+            if (!opt.nullable) {
+                defaultCast = (opt.nonEmpty
+                    ? util.castLowercaseNonEmptyString
+                    : util.castLowercaseString);
+            } else {
+                defaultCast = (opt.nonEmpty
+                    ? util.castLowercaseNonEmptyNullableString
+                    : util.castLowercaseNullableString);
+            }
+        } else if (type.includes('link')) {
+            if (!opt.nullable) {
+                defaultCast = util.castToRID;
+            } else {
+                defaultCast = util.castNullableLink;
+            }
         }
-        if ((prop.maxItems || prop.maxItems === 0) && result.length > prop.maxItems) {
-            throw new ValidationError(`Violated the maxItems constraint of ${prop.name}. More than the allowed number of elements (${result.length} > ${prop.maxItems})`);
+    }
+
+    let generateDefault,
```

Comment:
This one will have to wait, its hard to type it as is and i plan to change this a bit in the next PR so that the user doesn't pass default as generateDefault OR default at all but I don't want to make this PR any bigger than it already is.

partially addressed here: 425480b